src/Stratis.Bitcoin.Features.MemoryPool/Fee/BlockPolicyEstimator.cs

```diff
@@ -23,55 +27,84 @@ namespace Stratis.Bitcoin.Features.MemoryPool.Fee
-    /// stats on the transactions included in that block
+    /// When a transaction enters the mempool, we track the height of the block chain
+    /// at entry.  All further calculations are conducted only on this set of "seen"
+    /// transactions.Whenever a block comes in, we count the number of transactions
+    /// in each bucket and the total amount of feerate paid in each bucket. Then we
+    /// calculate how many blocks Y it took each transaction to be mined.  We convert
+    /// from a number of blocks to a number of periods Y' each encompassing "scale"
+    /// blocks.This is tracked in 3 different data sets each up to a maximum
+    /// number of periods.Within each data set we have an array of counters in each
+    /// feerate bucket and we increment all the counters from Y' up to max periods
+    /// representing that a tx was successfully confirmed in less than or equal to
+    /// that many periods. We want to save a history of this information, so at any
+    /// time we have a counter of the total number of transactions that happened in a
+    /// given feerate bucket and the total number that were confirmed in each of the
+    /// periods or less for any bucket.  We save this history by keeping an
+    /// exponentially decaying moving average of each one of these stats.  This is
+    /// done for a different decay in each of the 3 data sets to keep relevant data
+    /// from different time horizons.  Furthermore we also keep track of the number
+    /// unmined (in mempool or left mempool without being included in a block)
+    /// transactions in each bucket and for how many blocks they have been
+    /// outstanding and use both of these numbers to increase the number of transactions
+    /// we've seen in that feerate bucket when calculating an estimate for any number
+    /// of confirmations below the number of blocks they've been outstanding.
     /// </remarks>
     public class BlockPolicyEstimator
     {
-        /// <summary>Require an avg of 1 tx in the combined feerate bucket per block to have stat significance.</summary>
-        private const double SufficientFeeTxs = 1;
+        ///<summary>Track confirm delays up to 12 blocks for short horizon</summary>
+        private const int ShortBlockPeriods = 12;
+        private const int ShortScale = 1;
+        
+        ///<summary>Track confirm delays up to 48 blocks for medium horizon</summary>
+        private const int MedBlockPeriods = 24;
+        private const int MedScale = 2;
+
+        ///<summary>Track confirm delays up to 1008 blocks for long horizon</summary>
+        private const int LongBlockPeriods = 42;
+        private const int LongScale = 24;
+
+        ///<summary>Historical estimates that are older than this aren't valid</summary>
+        private const int OldestEstimateHistory = 6 * 1008;
+
+        ///<summary>Decay of .962 is a half-life of 18 blocks or about 3 hours</summary>
```

Comment:
All these comments suggest that this is bitcoin specific, at least the comments themselves are bitcoin specific it it is a question whether the same number of blocks should be used for other block chains.

Reply:
Yes, I see, I thought that horizons are rather counted in blocks, not in time intervals, so, e.g. medium horizon should be 48 blocks, no matter what is block time in concrete blockchain. I can suggest to make pair of settings for every horizon - for blocks and for minutes (or hours). E.g. MedBlocks and MedTime. If MedBlocks is filled - algorithm just gets its amount as block amount. If MedTime filled - algorithm gets time and divides it by block time for certain blockchain. It gives some flexibility in algorithm setup. Because honestly, I don't know what will be more correct - keep horizon duration the same as bitcoin has in blocks or in amount of time.

Reply:
About horizon decays - I can make them flexible by calculating them from block time (e.g. for bitcoin "Short history: 0.962 decay (half-life 18 blocks, 3 hours)", then for another blockchain it should be another number to suit this 3 hours)

Reply:
I'm not familiar with the algorithm yet, I just wanted to show you one example of what seems to be specific for bitcoin. You shouldn't then ask yourself if the same number should be used for blockchain which has different block size or different target spacing or different maximum number of transaction in a block or different recommended number of confirmations to wait for or different whatever.

If you realise that you really want to use the same number, then you just need to modify the comment appropriately so that it is not bitcoin specific. Otherwise you're looking for some kind of formula that will get you the correct number on each blockchain based on its parameters.